nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py

```diff
@@ -424,25 +431,95 @@ def loss_func(output_tensor):
+        # sanity checks
+        if len(args) != len(args_name):
+            raise ValueError(f"Mismatch between length in args_name ({len(args_name)}) and args ({len(args)})")
+        if any([n in kwargs for n in args_name]):
+            raise ValueError(f"args_name = {args_name} cannot overlap kwargs = {list(kwargs.keys())}")
+
+        # get mapping of kwarg names to arg index
+        kwargs_to_arg_idx = self._kwargs_to_arg_idx
+
+        # collect all arguments
+        all_args_name = args_name[:]
+        all_args = args[:]
+        for k, v in kwargs.items():
+            all_args_name.append(k)
+            all_args.append(v)
+
+        args_idx = [kwargs_to_arg_idx[n] for n in all_args_name]
+
+        # print(f"all_args_name = {all_args_name}   args_idx = {args_idx}")
+
+        # construct args ordered by name (with None as place-holder)
+        forward_args = [None] * (max(args_idx) + 1)
+        for i, v in zip(args_idx, all_args):
+            forward_args[i] = v
+
+        return forward_args
+
+    def get_forward_output_only_func_encode(self, arg_names, **kwargs):
+        """
+        args_idx - maps batch into index of args (with None filling gaps)
+        arg_names - corresponding names for a friendly error message
+        kwargs - shared arguments (non tensors)
+        """
+
         def fwd_output_only_func(batch, model):
             batch = [x.cuda(non_blocking=True) for x in batch]
-            if len(batch) == 4:
-                encoder_input_ids, decoder_input_ids, encoder_attn_mask, decoder_attn_mask = batch
-                enc_input = None
-            elif len(batch) == 5:
-                encoder_input_ids, decoder_input_ids, encoder_attn_mask, decoder_attn_mask, enc_input = batch
-            else:
-                raise ValueError("wrong number of items in the batch")
-            output = model(
-                encoder_input_ids,  # enc_input_ids
-                encoder_attn_mask,  # enc_attn_mask
-                decoder_input_ids,  # dec_input_ids
-                decoder_attn_mask,  # dec_attn_mask
-                None,  # token_type_ids
-                None,  # labels
-                enc_input,  # enc_hidden_states
-            )
+
+            # map batch and shared args into forward args
+            args = self._build_forward_args_from_kwargs(args_name=arg_names, args=batch, **kwargs)
+
+            # print(f"args = {args}")
+
+            output = model(*args)
+
+            def id_func(output_tensor):
+                return output_tensor, {'hiddens': output_tensor}
+
+            return output, id_func
+
+        return fwd_output_only_func
+
+    def get_forward_output_only_func_decode(self, arg_names, **kwargs):
+        """
+        args_idx - maps batch into index of args (with None filling gaps)
+        arg_names - corresponding names for a friendly error message
+        kwargs - shared arguments (non tensors)
+        """
+
+        def fwd_output_only_func(batch, model):
```

Comment:
Share with get_forward_output_only_func_encode, and add to docstring what are the outputs in id_func (e.g., 'logits')

Reply:
Not identical to decode. Let's discuss.